File: ios/HolyDay/Services/Common/ReminderPlanner.swift
//
//  ReminderPlanner.swift
//  HolyDay
//
//  Created by Matthias Cadet on 16/09/2026.
//

import Foundation

/// Ce que l'app sait de la vie de prière au moment où elle planifie les rappels. Les notifications
/// locales étant figées à la planification (et le store SwiftData illisible appareil verrouillé),
/// ce contexte est relu à chaque replanification, jamais au déclenchement.
struct ReminderContext: Equatable {
  var lastPrayerDate: Date?
  var lastEmotion: Emotion?
  var lastEmotionDate: Date?
  var oldestOpenIntentionDate: Date?

  static let empty = ReminderContext()
}

enum ReminderKind: Equatable {
  /// Question de réflexion tournante (comportement historique).
  case question
  /// Verset du thème de la dernière émotion déclarée ; index dans `VerseCorpus.all`.
  case verse(corpusIndex: Int)
  /// Invitation à revenir sur ses intentions — sans jamais en citer le texte (écran verrouillé).
  case intentions
}

/// Règles de contenu des rappels, pures et déterministes : miroir exact de `ReminderPlanner.kt`.
/// Toute modification doit être portée des deux côtés.
enum ReminderPlanner {
  /// Nombre de jours, après une prière portant une émotion, où le rappel propose un verset de ce
  /// thème. Court à dessein : on accompagne, on ne ressasse pas.
  static let emotionFollowUpDays = 2
  /// Âge minimal d'une intention ouverte avant qu'on invite à y revenir.
  static let intentionMinimumAgeDays = 7
  /// Dimanche (`Calendar.component(.weekday)`), jour fixe pour que l'invitation reste hebdomadaire
  /// même si l'app replanifie tous les jours.
  static let intentionWeekday = 1

  /// `nil` : pas de rappel ce jour-là (l'utilisateur a déjà prié — on ne relance pas).
  /// Priorité : silence > verset d'émotion > intentions du dimanche > question.
  static func kind(
    for fireDate: Date, context: ReminderContext, calendar: Calendar = .current
  ) -> ReminderKind? {
    if let lastPrayer = context.lastPrayerDate, calendar.isDate(lastPrayer, inSameDayAs: fireDate) {
      return nil
    }

    if let emotion = context.lastEmotion, let emotionDate = context.lastEmotionDate {
      let elapsed = days(from: emotionDate, to: fireDate, calendar: calendar)
      if (1...emotionFollowUpDays).contains(elapsed),
        let index = verseIndex(for: emotion, on: fireDate, calendar: calendar)
      {
        return .verse(corpusIndex: index)
      }
    }

    if calendar.component(.weekday, from: fireDate) == intentionWeekday,
      let oldest = context.oldestOpenIntentionDate,
      days(from: oldest, to: fireDate, calendar: calendar) >= intentionMinimumAgeDays
    {
      return .intentions
    }

    return .question
  }

  // Choix par jour de l'année et non via les decks mélangés de `VerseService` : planifier ne doit
  // pas consommer la pioche que l'utilisateur voit dans l'app.
  static func verseIndex(for emotion: Emotion, on date: Date, calendar: Calendar = .current)
    -> Int?
  {
    let corpus = VerseCorpus.all
    let pool = corpus.indices.filter { corpus[$0].emotionTags.contains(emotion.rawValue) }
    guard !pool.isEmpty else { return nil }
    let dayIndex = calendar.ordinality(of: .day, in: .year, for: date) ?? 1
    return pool[dayIndex % pool.count]
  }

  private static func days(from start: Date, to end: Date, calendar: Calendar) -> Int {
    let from = calendar.startOfDay(for: start)
    let to = calendar.startOfDay(for: end)
    return calendar.dateComponents([.day], from: from, to: to).day ?? 0
  }
}

File: android/app/src/main/java/com/matthiascadet/holyday/service/notification/ReminderPlanner.kt
package com.matthiascadet.holyday.service.notification

import com.matthiascadet.holyday.data.model.Emotion
import com.matthiascadet.holyday.data.model.VerseCorpus
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * Ce que l'app sait de la vie de prière au moment de choisir le contenu d'un rappel. Sur Android
 * il est lu au déclenchement (le récepteur tourne dans le process de l'app), là où iOS doit le
 * figer à la planification — le résultat perçu est le même.
 */
data class ReminderContext(
    val lastPrayerDate: LocalDate? = null,
    val lastEmotion: Emotion? = null,
    val lastEmotionDate: LocalDate? = null,
    val oldestOpenIntentionDate: LocalDate? = null,
)

sealed interface ReminderKind {
    /** Question de réflexion tournante (comportement historique). */
    data object Question : ReminderKind

    /** Verset du thème de la dernière émotion déclarée ; index dans `VerseCorpus.all`. */
    data class Verse(val corpusIndex: Int) : ReminderKind

    /** Invitation à revenir sur ses intentions — sans jamais en citer le texte (écran verrouillé). */
    data object Intentions : ReminderKind
}

/**
 * Règles de contenu des rappels, pures et déterministes : miroir exact de `ReminderPlanner.swift`.
 * Toute modification doit être portée des deux côtés.
 */
object ReminderPlanner {
    /** Jours de suivi après une prière portant une émotion. Court à dessein : on accompagne, on ne ressasse pas. */
    const val EMOTION_FOLLOW_UP_DAYS = 2L

    /** Âge minimal d'une intention ouverte avant qu'on invite à y revenir. */
    const val INTENTION_MINIMUM_AGE_DAYS = 7L

    /** Jour fixe, pour que l'invitation reste hebdomadaire. */
    val INTENTION_DAY: DayOfWeek = DayOfWeek.SUNDAY

    /**
     * `null` : pas de rappel ce jour-là (l'utilisateur a déjà prié — on ne relance pas).
     * Priorité : silence > verset d'émotion > intentions du dimanche > question.
     */
    fun kindFor(day: LocalDate, context: ReminderContext): ReminderKind? {
        if (context.lastPrayerDate == day) return null

        val emotion = context.lastEmotion
        val emotionDate = context.lastEmotionDate
        if (emotion != null && emotionDate != null) {
            val elapsed = ChronoUnit.DAYS.between(emotionDate, day)
            if (elapsed in 1..EMOTION_FOLLOW_UP_DAYS) {
                verseIndex(emotion, day)?.let { return ReminderKind.Verse(it) }
            }
        }

        val oldest = context.oldestOpenIntentionDate
        if (day.dayOfWeek == INTENTION_DAY && oldest != null &&
            ChronoUnit.DAYS.between(oldest, day) >= INTENTION_MINIMUM_AGE_DAYS
        ) {
            return ReminderKind.Intentions
        }

        return ReminderKind.Question
    }

    /**
     * Choix par jour de l'année et non via les pioches de `VerseService` : préparer un rappel ne
     * doit pas consommer la pioche que l'utilisateur voit dans l'app.
     */
    fun verseIndex(emotion: Emotion, day: LocalDate): Int? {
        val pool = VerseCorpus.all.indices.filter { VerseCorpus.all[it].emotionTags.contains(emotion.id) }
        if (pool.isEmpty()) return null
        return pool[day.dayOfYear % pool.size]
    }
}
